frontend/www/js/omegaup/arena/events_socket.ts

```diff
@@ -0,0 +1,218 @@
+      this.locationHost = options.locationHost;
+      this.problemsetId = options.problemsetId;
+      this.scoreboardToken = options.scoreboardToken;
+      this.clarificationsOffset = options.clarificationsOffset;
+      this.clarificationsRowcount = options.clarificationsRowcount;
+    }
+  }
+
+  connect(): Promise<void> {
+    this.shouldRetry = false;
+    return new Promise((accept, reject) => {
+      try {
+        const socket = new WebSocket(this.uri, 'com.omegaup.events');
+
+        socket.onmessage = (message) => this.onmessage(message);
+        socket.onopen = () => {
+          this.shouldRetry = true;
+          this.socketStatus = SocketStatus.Ok;
+          this.socketKeepalive = setInterval(
+            () => socket.send('"ping"'),
+            30000,
+          );
+          accept();
+        };
+        socket.onclose = (e: Event) => {
+          this.onclose();
+          reject(e);
+        };
+
+        this.socket = socket;
+      } catch (e) {
+        reject(e);
+        return;
+      }
+    });
+  }
+
+  onmessage(message: MessageEvent) {
+    const data = JSON.parse(message.data);
+
+    if (data.message == '/run/update/') {
+      data.run.time = time.remoteTime(data.run.time * 1000);
+      updateRun(data.run);
+    } else if (data.message == '/clarification/update/') {
+      data.clarification.time = time.remoteTime(data.clarification.time * 1000);
+      clarificationStore.commit('addClarification', data.clarification);
+    } else if (data.message == '/scoreboard/update/') {
+      data.time = time.remoteTime(data.time * 1000);
+      // TODO: Uncomment next block when virtual contest is migrated
+      /*if (problemsetAdmin && data.scoreboard_type != 'admin') {
+        if (options.originalContestAlias == null) return;
+        virtualRankingChange(data.scoreboard);
+        return;
+      }*/
+      onRankingChanged({ scoreboard: data.scoreboard });
+
+      api.Problemset.scoreboardEvents({
+        problemset_id: this.problemsetId,
+        token: this.scoreboardToken,
+      })
+        .then((response) => onRankingEvents(response.events))
+        .catch(ui.ignoreError);
+    }
+  }
+
+  onclose() {
+    this.socket = null;
+    if (this.socketKeepalive) {
+      clearInterval(this.socketKeepalive);
+      this.socketKeepalive = null;
+    }
+    if (this.shouldRetry && this.retries > 0) {
+      this.retries--;
+      this.socketStatus = SocketStatus.Waiting;
+      setTimeout(() => this.connect(), Math.random() * 15000);
+      return;
+    }
+    this.socketStatus = SocketStatus.Failed;
+  }
+
+  connectSocket(): void {
+    if (this.disableSockets || this.problemsetAlias == 'admin') {
+      this.socketStatus = SocketStatus.Failed;
+      return;
+    }
+
+    this.uri =
+      `${window.location.protocol === 'https:' ? 'wss:' : 'ws:'}//${
+        window.location.host
+      }/events/?filter=/problemset/${this.problemsetId}` +
+      (this.scoreboardToken ? `/${this.scoreboardToken}` : '');
```

Comment:
esto se puede establecer en el constructor. así no se usa `''` como valor mágico en `this.uri`. y además, esto debería usar las propiedades:

```suggestion
    this.uri =
      `${this.locationProtocol === 'https:' ? 'wss:' : 'ws:'}//${
        this.locationHost
      }/events/?filter=/problemset/${this.problemsetId}` +
      (this.scoreboardToken ? `/${this.scoreboardToken}` : '');
```

Reply:
Listo.